### DISCUSSION TITLE
Pessimistic concurrency control

### DISCUSSION THREAD

**Discussion by finchi** [bensheldon/good_job#840] [Q&A]
I have an application where customers can send multiple (customized) API requests to the same endpoint. They are predefined once and run every 5 minutes afterwards. The API endpoint is rate limited per customer, and I have no control over it.
Which basically comes down to, only one request (job) per customer is allowed to run at the same time.

What I wanted to do originally, was just put every API request as a job into the queue, with a `concurrency_key` from the customer and a `perform_limit` of 1. I quickly learned that `good_job` expects am optimistic approach to concurrency, and that this method is not working for our case (far too many reschedules because of hitting the `perform_limit`).

Next, I was thinking about the approach described in the documentation, having one queue per customer and setting up a thread for each customer. Put apart from not being very flexible, changing the config with every new customer and restarting the service, it’s also not horizontally scalable. 

**What we’re currently doing**: we have 1 cronjob that queries over all requests per customer, order by last performed, and only queue the oldest request for `good_job`. This is working fine, but leaves gaps in-between jobs since I only can queue another job after one is finished.

I just wanted to ask is there is something I’m missing how this can be done, or if there is something different I could do with `good_job`

Thank you very much for this awesome gem!

**Comment by bensheldon** [comment]
Would it be possible for a single job to emit multiple API requests? I'm imagining something like this, which you would _attempt_ to enqueue every time an ApiRequest object is created, but the `total_limit` would ensure that only one job ever existed at once.

```ruby
class CustomerApiRequestsJob < ApplicationJob
  good_job_control_concurrency_with total_limit: 1, key: -> { "customer-#{args.first.id}" }

  def perform(customer)
    # continually emit API requests until there are no more
    while api_request = ApiRequests.where(customer: customer).order(created_at: :asc).where("quarantined_until IS NULL OR quarantined_until < ?", Time.current).first
      ApiRequest.send_it
      ApiRequest.destroy
    rescue StandardError
      ApiRequest.update(quarantined_until: 5.minutes.from_now)
    end
  end
end
```

Another idea, if you wanted to maintain "one job per API request" is to put each job onto a custom queue (`"customer-#{customer.id}"`)for each customer that you don't provision any processes/threads to work from. You could then have a similar job to the above, but instead of the `while`-loop, you called `GoodJob.perform_inline("customer-#{customer.id}")` which would run the jobs inline inside of the job.
